platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy

```diff
@@ -0,0 +1,144 @@
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.security.idp.plugin.audiencerestriction
+
+import ddf.security.samlp.SamlProtocol
+import org.apache.wss4j.common.saml.OpenSAMLUtil
+import org.opensaml.saml.saml2.core.Assertion
+import org.opensaml.saml.saml2.core.Audience
+import org.opensaml.saml.saml2.core.AudienceRestriction
+import org.opensaml.saml.saml2.core.AuthnRequest
+import org.opensaml.saml.saml2.core.Conditions
+import org.opensaml.saml.saml2.core.Issuer
+import org.opensaml.saml.saml2.core.Response
+import spock.lang.Specification
+
+class AudienceRestrictionPresignPluginSpec extends Specification {
+    def ISSUER_ID = "https://localhost:8993/services/idp/login"
+
+    AudienceRestrictionPresignPlugin plugin = new AudienceRestrictionPresignPlugin()
+
+    // general mocks
+    AuthnRequest authNReq = Mock(AuthnRequest)
+    Issuer issuer = Mock(Issuer)
+
+    List<String> spMetadata = new ArrayList<>()
+    Set<SamlProtocol.Binding> bindings = new HashSet<>()
+
+    // heterogenous set of mocks and concrete objects
+    Response response = Mock(Response)
+    List<Assertion> assertions = new ArrayList<>()
+    Assertion assertion = Mock(Assertion)
+    Conditions conditions = Mock(Conditions)
+    List<AudienceRestriction> audienceRestrictions = new ArrayList<>()
+
+    AudienceRestriction audienceRestriction = Mock(AudienceRestriction)
+    AudienceRestriction audienceRestriction2 = Mock(AudienceRestriction)
+    AudienceRestriction emptyAudienceRestriction = Mock(AudienceRestriction)
+
+    List<Audience> audiences = new ArrayList<>()
+    List<Audience> audiences2 = new ArrayList<>()
+    List<Audience> emptyAudiences = new ArrayList<>()
+
+    Audience audience = Mock(Audience)
+    Audience audience2 = Mock(Audience)
+
+    def setupSpec() {
+        OpenSAMLUtil.initSamlEngine()
+    }
+
+    def setup() {
+        // set up authNReq
+        authNReq.getIssuer() >> issuer
+        issuer.getValue() >> ISSUER_ID
+
+        // set up basic response
+        response.assertions >> assertions
+        assertions.add(assertion)
+        assertion.conditions >> conditions
+        conditions.audienceRestrictions >> audienceRestrictions
+
+        // set up audienceURIs
+        audience.audienceURI >> "https://localhost:8993/services/SecurityTokenService?wsdl"
+        audience2.audienceURI >> "https://localhost:8993/admin"
+
+        // add the audience to the audiences
+        audiences.add(audience)
+        audiences.add(audience2)
+        audiences2.add(audience)
+
+        // stub the audiences to the audienceRestriction
+        audienceRestriction.audiences >> audiences
+        audienceRestriction2.audiences >> audiences2
+        emptyAudienceRestriction.audiences >> emptyAudiences
+    }
+
+    def 'add audience to non existent audience restriction'() {
```

Comment:
❓ To an empty audience restriction or a non-existing?

Reply:
Non-existent

Reply:
It's not clear where you're doing that. Is that all the things you did in setup?

Reply:
I did do it in setup. Tried to clear it up with a comment.